server/game/gameSystems/DelayedEffectSystem.ts

```diff
@@ -0,0 +1,71 @@
+import AbilityHelper from '../AbilityHelper';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import { IAbilityLimit } from '../core/ability/AbilityLimit';
+import { TriggeredAbilityContext } from '../core/ability/TriggeredAbilityContext';
+import { Duration, EffectName, EventName, PhaseName } from '../core/Constants';
+import { GameEvent } from '../core/event/GameEvent';
+import { GameSystem, IGameSystemProperties } from '../core/gameSystem/GameSystem';
+import { OngoingEffectBuilder } from '../core/ongoingEffect/OngoingEffectBuilder';
+import { WhenType } from '../Interfaces';
+
+export interface IDelayedEffectSystemProperties extends IGameSystemProperties {
+    when: WhenType;
+    duration: Duration;
+    limit?: IAbilityLimit;
+    immediateEffect: GameSystem<TriggeredAbilityContext>;
+}
+
+export class DelayedEffectSystem<TContext extends AbilityContext = AbilityContext> extends GameSystem<TContext, IDelayedEffectSystemProperties> {
+    public override readonly name: string = 'applyDelayedEffect';
+    public override readonly eventName: EventName = EventName.OnEffectApplied;
+    public override readonly effectDescription: string = 'apply a delayed effect';
+
+    protected override defaultProperties: IDelayedEffectSystemProperties = {
+        when: null,
+        duration: null,
+        limit: AbilityHelper.limit.perGame(1),
+        immediateEffect: null
+    };
+
+    public eventHandler(event: any, additionalProperties: any): void {
+        const properties = this.generatePropertiesFromContext(event.context, additionalProperties);
+        // TODO Remove this if we don't need it
+        // if (!properties.ability) {
+        //     properties.ability = event.context.ability;
+        // }
+
+        const { when, duration, limit, immediateEffect, ...otherProperties } = properties;
+
+        const delayedEffect = AbilityHelper.ongoingEffects.delayedEffect({
+            title: 'Discard a card from your hand',
+            when: when,
+            immediateEffect: immediateEffect,
+            limit: limit
+        });
+
+        const renamedProperties = Object.assign(otherProperties, { ongoingEffect: delayedEffect });
+
+        event.context.source[duration](() => renamedProperties);
```

Comment:
Looks like player target lasting effects and card target lasting effects do this differently. The card version is explicitly registering with the OngoingEffectEngine but the player version is doing this, which is calling a method on `OngoingEffectSource` based on the duration.

As an aside, probably this should be something more explicit like a switch since there are some values of Duration that don't have a corresponding method (and shouldn't), and also b/c this is impossible to read. Took me a while to piece together what it was doing.